### ISSUE TITLE
MongoDB connection pooling leaks connections

### ISSUE THREAD

**Issue by aticu** [SergioBenitez/Rocket#979]
# Bug report

## What is the bug?

When using the MongoDB connection pooling feature, Rocket opens new connections to the MongoDB server on a regular basis (roughly every 11 minutes), leaking the old connections. This happens both in production and development mode. It also happens regardless of any requests being made.

## Version information

Rocket version: 0.4.0
Operating system: Tested on a fresh install of Ubuntu 18.04.1 LTS

### Setup

`Cargo.toml`

```toml
[package]
name = "package"
version = "0.1.0"
authors = ["aticu"]
edition = "2018"

[dependencies]
rocket = "0.4"
rocket_contrib = { version = "0.4", features = ["mongodb_pool"] }
```

`src/main.rs`

```rust
#[rocket_contrib::database("db")]
struct DatabaseConnection(rocket_contrib::databases::mongodb::db::Database);

fn main() {
    rocket::ignite()
        .attach(DatabaseConnection::fairing())
        .launch();
}
```

`Rocket.toml`

```toml
[global.databases]
db = { url = "mongodb://localhost:27017" }
```

### Steps

- Start mongodb and run `cargo run`
- Check the number of open connections with `lsof -p $(pidof package) | grep localhost | grep -v LISTEN | wc -l`
- Wait roughly 11 minutes
- Check the number of connections again: It should have doubled

# Edit

It seems as though there are a lot more open connections than I initially thought. If you use `lsof | grep package | grep localhost | grep -v LISTEN | wc -l` you can see the actual number of open connections opened by the program. Most of those seem to be opened by a thread named `r2d2-work`. This number seems to increase in the same 11 minute interval.

**Comment by jebrosen** [comment]
Rocket doesn't do anything more than set up an r2d2 pool, so this issue is more likely in r2d2-mongodb or mongodb. 11 minutes sounds familiar; I think r2d2 tries to keep "fresh" connections every 10 minutes or so. I'll try to reproduce this and see if I have any insights.

**Comment by schra** [comment]
@jebrosen Got any insights? Are you certain that the issue lies in rd2d and not in rocket? If you are certain about this, I'll create an upstream issue for this. Although I wouldn't be able to provide a MWE for the upstream issue since I'm not familiar enough with rocket's source code.

What I found so far is this function https://github.com/sfackler/r2d2/blob/dd390f674543c03b12ced993f704f617fc221421/src/config.rs#L157 where the default idle time is 10 minutes (just as you said). Not sure if the idle time is the problem, but if it is, could rocket maybe just use no idle time as a workaround until they fixed this upstream?

**Comment by jebrosen** [comment]
I still have not gotten around to reproducing it, sorry. But it shouldn't be too hard for anyone interested to test: a short program that creates an r2d2 Pool and then doesn't exit (sleep in a loop, for example) should have the same problem if it's in the r2d2 wrapper or mongodb. If it's in mongodb itself, creating and destroying connections repeatedly in the same way that the r2d2 wrapper crates them would also have the same problem.

**Comment by aticu** [comment]
I was able to reproduce the issue with just `r2d2-mongodb` and opened up an issue upstream:

https://gitlab.com/petoknm/r2d2-mongodb/issues/4

**Comment by jebrosen** [comment]
The underlying issue appears to be mongodb-labs/mongo-rust-driver-prototype#247.

**Comment by Liby99** [comment]
My Rocket server can have a memory usage easily going up to 900+M while my old NodeJs version will consistently use 100+M... This is pretty annoying and the memory leak prevents me from putting it onto production.

**Comment by SergioBenitez** [comment]
@Liby99 The issue you're experiencing isn't caused by software we control. If you'd like a resolution, I suggest you create an issue in the repository @jebrosen linked to.

**Comment by zhhzhhz** [comment]
The reason for this problem is that the socket connection does not break when the mongodb client is released.
r2d2_mongodb use 'idle_timeout' and 'max_lifetime' to check which connection should be release and reconnect.
so you can fix this by use "r2d2::Pool::builder().max_size(16).idle_timeout(None).max_lifetime(None).build(manager) ", and you will see twice as many sockets as max_size.

**Comment by zhhzhhz** [comment]
> I was able to reproduce the issue with just `r2d2-mongodb` and opened up an issue upstream:
> 
> https://gitlab.com/petoknm/r2d2-mongodb/issues/4

The reason for this problem is that the socket connection does not break when the mongodb client is released.
r2d2_mongodb use 'idle_timeout' and 'max_lifetime' to check which connection should be release and reconnect.
so you can fix this by use "r2d2::Pool::builder().max_size(16).idle_timeout(None).max_lifetime(None).build(manager) ", and you will see twice as many sockets as max_size.

**Comment by AshtonKem** [comment]
This issue should be solved in r2d2-mongodb 0.3.0, which means that this ticket can be closed.

**Comment by SergioBenitez** [comment]
What's more, `master` no longer includes `mongodb`. Closing.
